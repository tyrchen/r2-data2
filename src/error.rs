use axum::{
    Json,
    http::StatusCode,
    response::{IntoResponse, Response},
};
use serde_json::json;
use thiserror::Error;
use tracing::warn;

#[derive(Error, Debug)]
pub enum AuthError {
    #[error("Invalid token: {0}")]
    InvalidToken(String),

    #[error("Missing credentials")]
    MissingCredentials,

    #[error("Token creation error")]
    TokenCreation,

    #[error("Internal server error (auth)")]
    InternalError,
}

// General AppError Enum
#[derive(Error, Debug)]
pub enum AppError {
    #[error(transparent)]
    Auth(#[from] AuthError),

    #[error("Database error: {0}")]
    Database(#[from] sqlx::Error),

    #[error("Unsupported database type: {0}")]
    UnsupportedDatabaseType(String),

    #[error("Configuration error: {0}")]
    Config(#[from] config::ConfigError),

    #[error("Not found: {0}")]
    NotFound(String),

    #[error("Not implemented: {0}")]
    NotImplemented(String),

    #[error("Bad request: {0}")]
    BadRequest(String),

    #[error("SQL parsing error: {0}")]
    SqlParsingError(String),

    #[error("Invalid query result: {0}")]
    InvalidQueryResult(String),
}

impl IntoResponse for AuthError {
    fn into_response(self) -> Response {
        let (status, error_message) = match self {
            AuthError::InvalidToken(msg) => {
                (StatusCode::UNAUTHORIZED, format!("Invalid Token: {}", msg))
            }
            AuthError::MissingCredentials => {
                (StatusCode::UNAUTHORIZED, "Missing Credentials".to_string())
            }
            AuthError::TokenCreation => (
                StatusCode::INTERNAL_SERVER_ERROR,
                "Internal error".to_string(),
            ),
            AuthError::InternalError => (
                StatusCode::INTERNAL_SERVER_ERROR,
                "Internal error".to_string(),
            ),
        };

        let body = Json(json!({ "error": error_message }));
        (status, body).into_response()
    }
}

impl IntoResponse for AppError {
    fn into_response(self) -> Response {
        let (status, error_message) = match self {
            AppError::Auth(auth_error) => {
                // Reuse AuthError's IntoResponse implementation detail logic
                return auth_error.into_response();
            }
            AppError::Database(db_err) => {
                // Log the detailed DB error for internal review
                tracing::error!("Database error: {}", db_err);
                // Provide a generic message to the client
                (
                    StatusCode::INTERNAL_SERVER_ERROR,
                    "Internal database error".to_string(),
                )
            }
            AppError::UnsupportedDatabaseType(db_type) => (
                StatusCode::BAD_REQUEST,
                format!("Unsupported database type: {}", db_type),
            ),
            AppError::Config(config_err) => {
                tracing::error!("Configuration error: {}", config_err);
                (
                    StatusCode::INTERNAL_SERVER_ERROR,
                    "Server configuration error".to_string(),
                )
            }
            AppError::NotFound(msg) => (StatusCode::NOT_FOUND, msg),
            AppError::NotImplemented(msg) => (StatusCode::NOT_IMPLEMENTED, msg),
            AppError::BadRequest(msg) => (StatusCode::BAD_REQUEST, msg),
            AppError::SqlParsingError(msg) => (StatusCode::BAD_REQUEST, msg),
            AppError::InvalidQueryResult(msg) => {
                warn!("Invalid query result: {}", msg);
                (
                    StatusCode::INTERNAL_SERVER_ERROR,
                    "Invalid query result".to_string(),
                )
            }
        };

        let body = Json(json!({ "error": error_message }));
        (status, body).into_response()
    }
}
