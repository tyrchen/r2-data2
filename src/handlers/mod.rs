use crate::{
    db::{DatabaseInfo, PoolHandler, TableInfo, TableSchema},
    error::AppError,
    state::AppState,
};
use axum::{
    Json,
    extract::{Path, State},
};
use serde::Deserialize;
use serde_json::{Value, json};

#[derive(Deserialize)]
pub struct ExecuteQueryRequest {
    pub db_name: String,
    pub query: String,
}

// Placeholder handler for authenticated routes
pub async fn ping() -> Json<Value> {
    Json(json!({ "message": "pong" }))
}

// Handler to list configured databases
pub async fn list_databases(State(state): State<AppState>) -> Json<Vec<DatabaseInfo>> {
    let databases_info: Vec<DatabaseInfo> = state
        .config
        .databases
        .iter()
        .map(|db_config| DatabaseInfo {
            name: db_config.name.clone(),
            db_type: db_config.db_type.to_string(), // Convert enum to string
        })
        .collect();

    Json(databases_info)
}

pub async fn list_tables(
    State(state): State<AppState>,
    Path(db_name): Path<String>,
) -> Result<Json<Vec<TableInfo>>, AppError> {
    // Directly access the pool via the Arc'd HashMap
    // Papaya hashmap is designed for concurrent reads
    let pools = state.pools.pin_owned();
    let pool = pools
        .get(&db_name)
        .ok_or_else(|| AppError::NotFound(format!("Database '{}' not found", db_name)))?;

    let tables = pool.list_tables().await?;
    Ok(Json(tables))
}

pub async fn get_table_schema(
    State(state): State<AppState>,
    Path((db_name, table_name)): Path<(String, String)>,
) -> Result<Json<TableSchema>, AppError> {
    let pools = state.pools.pin_owned();
    let pool = pools
        .get(&db_name)
        .ok_or_else(|| AppError::NotFound(format!("Database '{}' not found", db_name)))?;

    // Call the abstracted method on the pool
    let schema = pool.get_table_schema(&table_name).await?;

    Ok(Json(schema))
}

pub async fn execute_query(
    State(state): State<AppState>,
    Json(payload): Json<ExecuteQueryRequest>,
) -> Result<Json<Value>, AppError> {
    let db_name = payload.db_name;
    let pools = state.pools.pin_owned();
    let pool = pools
        .get(&db_name)
        .ok_or_else(|| AppError::NotFound(format!("Database '{}' not found", db_name)))?;

    let sql = pool.sanitize_query(&payload.query).await?;
    let data = pool.execute_query(&sql).await?;
    Ok(Json(data))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{AppConfig, TableType, db::ColumnType, state::AppState};
    use axum::{Json, extract::State};

    #[derive(Deserialize)]
    struct User {
        id: i32,
        name: String,
        email: String,
        #[allow(dead_code)]
        password: String,
    }

    #[tokio::test]
    async fn test_list_databases() {
        let state = AppState::new(AppConfig::load().unwrap()).await.unwrap();
        let Json(response) = list_databases(State(state)).await;
        assert_eq!(response.len(), 1);
        assert_eq!(response[0].name, "users");
        assert_eq!(response[0].db_type, "postgres");
    }

    #[tokio::test]
    async fn test_list_tables() {
        let state = AppState::new(AppConfig::load().unwrap()).await.unwrap();
        let Json(response) = list_tables(State(state), Path("users".to_string()))
            .await
            .unwrap();
        println!("response: {:?}", response);
        assert_eq!(response.len(), 5);
        assert_eq!(response[0].name, "public.repositories");
        assert_eq!(response[0].table_type, TableType::Table);
    }

    #[tokio::test]
    async fn test_get_table_schema() {
        let state = AppState::new(AppConfig::load().unwrap()).await.unwrap();
        let Json(response) = get_table_schema(
            State(state),
            Path(("users".to_string(), "repository_members".to_string())),
        )
        .await
        .unwrap();
        assert_eq!(response.columns.len(), 3);
        assert_eq!(response.columns[0].name, "id");
        assert_eq!(response.columns[0].data_type, ColumnType::Integer);
        assert!(!response.columns[0].is_nullable);
        assert!(response.columns[0].is_pk);
        assert!(response.columns[0].is_unique);

        assert_eq!(response.columns[1].name, "repository_id");
        assert_eq!(response.columns[1].data_type, ColumnType::Integer);
        assert!(!response.columns[1].is_nullable);
        assert!(!response.columns[1].is_pk);
        assert!(response.columns[1].is_unique);
        assert_eq!(
            response.columns[1].fk_table,
            Some("repositories".to_string())
        );
        assert_eq!(response.columns[1].fk_column, Some("id".to_string()));

        assert_eq!(response.columns[2].name, "user_id");
        assert_eq!(response.columns[2].data_type, ColumnType::Integer);
        assert!(!response.columns[2].is_nullable);
        assert!(!response.columns[2].is_pk);
        assert!(response.columns[2].is_unique);
        assert_eq!(response.columns[2].fk_table, Some("users".to_string()));
        assert_eq!(response.columns[2].fk_column, Some("id".to_string()));
    }

    #[tokio::test]
    async fn test_execute_query() {
        let state = AppState::new(AppConfig::load().unwrap()).await.unwrap();
        let Json(data) = execute_query(
            State(state),
            Json(ExecuteQueryRequest {
                db_name: "users".to_string(),
                query: "SELECT * FROM users".to_string(),
            }),
        )
        .await
        .unwrap();
        println!("data: {:?}", data);
        let users: Vec<User> = serde_json::from_value(data).unwrap();
        assert_eq!(users[0].id, 1);
        assert_eq!(users[0].name, "Alice Johnson");
        assert_eq!(users[0].email, "alice@example.com");
    }
}
