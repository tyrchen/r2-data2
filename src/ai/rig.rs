use crate::error::AppError;
use crate::handlers::FullSchema;
use rig::OneOrMany;
use rig::completion::Chat;
use rig::message::Message;
use rig::message::{AssistantContent, UserContent};
use rig::providers::openai as rig_openai;
use tracing::{error, info, instrument};

// Placeholder for the AI query generation logic
#[instrument(skip(openai_client, schema), fields(db_name = %db_name))]
pub async fn generate_sql_query(
    openai_client: &rig_openai::Client,
    db_name: &str,
    schema: &FullSchema, // Or maybe just DatabaseSchema?
    prompt: &str,
) -> Result<String, AppError> {
    info!("Generating SQL query using AI for database: {}", db_name);

    // TODO: 1. Format the schema into a string (e.g., Markdown)
    let schema_string = format_schema_for_prompt(schema, db_name)?;

    // Construct the prompt using rig::completion::Prompt
    // System prompt provides context and instructions
    let system_prompt = format!(
        r#"You are an expert SQL assistant. You are connected to a database named '{}'.
        Given the following database schema (in Markdown format), write a single, valid SQL query
        that precisely answers the user's request. Only output the pure SQL query, no code fence, no backticks, no additional explanation or text.
        "\n\nDatabase Schema:\n```markdown\n{}\n```"#,
        db_name, schema_string
    );

    // User prompt contains the specific request
    let user_prompt = prompt.to_string();

    // Define the model to use (e.g., gpt-4o)
    let model = "gpt-4o";
    info!("Prompting model '{}'", model);

    // Build the agent and send the prompt
    let agent = openai_client.agent(model).build();

    // Construct messages for the chat API
    let messages = vec![Message::Assistant {
        content: OneOrMany::one(AssistantContent::Text(system_prompt.into())),
    }];

    let prompt = Message::User {
        content: OneOrMany::one(UserContent::Text(user_prompt.into())),
    };

    match agent.chat(prompt, messages).await {
        Ok(response) => {
            info!("Successfully received response from AI model.");
            if response.is_empty() {
                error!("AI returned an empty response.");
                return Err(AppError::AiError(
                    "AI returned an empty response.".to_string(),
                ));
            }

            info!("Generated SQL query: {}", response);
            Ok(response)
        }
        Err(e) => {
            error!("Error calling OpenAI API: {}", e);
            // Convert rig::Error into AppError::AiError
            Err(AppError::AiError(format!(
                "Failed to generate query: {}",
                e
            )))
        }
    }
}

// Placeholder for schema formatting logic
fn format_schema_for_prompt(schema: &FullSchema, db_name: &str) -> Result<String, AppError> {
    // Find the specific database schema
    let db_schema = schema
        .databases
        .iter()
        .find(|db| db.name == db_name)
        .ok_or_else(|| AppError::NotFound(format!("Schema not found for database: {}", db_name)))?;

    // Simple Markdown formatting (can be enhanced)
    let mut markdown = format!("# Database: {}\n\n", db_schema.name);
    for table in &db_schema.tables {
        markdown.push_str(&format!("## Table: {}\n", table.table_name));
        markdown.push_str("| Column | Type | Nullable | PK | FK |\n");
        markdown.push_str("|---|---|---|---|---|\n");
        for col in &table.columns {
            markdown.push_str(&format!(
                "| {} | {:?} | {} | {} | {} |\n",
                col.name,
                col.data_type,
                if col.is_nullable { "YES" } else { "NO" },
                if col.is_pk { "YES" } else { "NO" },
                col.fk_table.as_ref().map_or("NO".to_string(), |t| format!(
                    "-> {}.{}",
                    t,
                    col.fk_column.as_deref().unwrap_or("?")
                ))
            ));
        }
        markdown.push('\n');
    }

    Ok(markdown)
}

#[cfg(test)]
mod tests {
    use super::*;
    // Import necessary structs directly from db and handlers
    use crate::db::{ColumnInfo, ColumnType, TableSchema}; // Import directly from db
    use crate::handlers::{DatabaseSchema, FullSchema};
    use insta::assert_snapshot;

    #[test]
    fn test_format_schema_simple() {
        // Arrange: Create mock schema data
        let db_schema = DatabaseSchema {
            name: "test_db".to_string(),
            db_type: "postgresql".to_string(),
            tables: vec![
                TableSchema {
                    table_name: "users".to_string(),
                    columns: vec![
                        ColumnInfo {
                            name: "id".to_string(),
                            data_type: ColumnType::Other("integer".to_string()), // Wrap type
                            is_nullable: false,
                            is_pk: true,
                            is_unique: false,
                            fk_table: None,
                            fk_column: None,
                        },
                        ColumnInfo {
                            name: "username".to_string(),
                            data_type: ColumnType::Other("text".to_string()), // Wrap type
                            is_nullable: false,
                            is_pk: false,
                            is_unique: true,
                            fk_table: None,
                            fk_column: None,
                        },
                    ],
                },
                TableSchema {
                    table_name: "posts".to_string(),
                    columns: vec![
                        ColumnInfo {
                            name: "post_id".to_string(),
                            data_type: ColumnType::Other("integer".to_string()), // Wrap type
                            is_nullable: false,
                            is_pk: true,
                            is_unique: false,
                            fk_table: None,
                            fk_column: None,
                        },
                        ColumnInfo {
                            name: "user_id".to_string(),
                            data_type: ColumnType::Other("integer".to_string()), // Wrap type
                            is_nullable: false,
                            is_pk: false,
                            is_unique: false,
                            fk_table: Some("users".to_string()),
                            fk_column: Some("id".to_string()),
                        },
                        ColumnInfo {
                            name: "content".to_string(),
                            data_type: ColumnType::Other("text".to_string()), // Wrap type
                            is_nullable: true,
                            is_pk: false,
                            is_unique: false,
                            fk_table: None,
                            fk_column: None,
                        },
                    ],
                },
            ],
        };
        let full_schema = FullSchema {
            databases: vec![db_schema],
        };

        // Act: Call the function
        let result = format_schema_for_prompt(&full_schema, "test_db");

        // Assert: Check if successful and compare with snapshot
        assert!(result.is_ok());
        assert_snapshot!(result.unwrap());
    }

    #[test]
    fn test_format_schema_db_not_found() {
        // Arrange: Empty schema
        let full_schema = FullSchema { databases: vec![] };

        // Act: Call the function with a non-existent db name
        let result = format_schema_for_prompt(&full_schema, "non_existent_db");

        // Assert: Check for NotFound error
        assert!(result.is_err());
        match result.err().unwrap() {
            AppError::NotFound(msg) => {
                assert!(msg.contains("Schema not found for database: non_existent_db"));
            }
            _ => panic!("Expected NotFound error"),
        }
    }
}
