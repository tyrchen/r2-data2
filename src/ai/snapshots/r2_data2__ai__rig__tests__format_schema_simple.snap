---
source: src/ai/rig.rs
expression: result.unwrap()
---
# Database: test_db

## Table: users
| Column | Type | Nullable | PK | FK |
|---|---|---|---|---|
| id | Other("integer") | NO | YES | NO |
| username | Other("text") | NO | NO | NO |

## Table: posts
| Column | Type | Nullable | PK | FK |
|---|---|---|---|---|
| post_id | Other("integer") | NO | YES | NO |
| user_id | Other("integer") | NO | NO | -> users.id |
| content | Other("text") | YES | NO | NO |
